### PR TITLE
improve documentation how OS is looked up

### LINKED ISSUE #946: updating an os with the operatingsystem module always fails
##### SUMMARY
When using the operatingsystem module you get an error when you try to update an operating system in katello
Adding a new operating system to katello works fine.

##### ISSUE TYPE
 - Bug Report

##### ANSIBLE VERSION

```
ansible 2.9.11
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/rob/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.5 (default, Aug 12 2020, 00:00:00) [GCC 10.2.1 20200723 (Red Hat 10.2.1-1)]
```

##### KATELLO/FOREMAN VERSION

```
tfm-rubygem-katello-3.16.0-1.el7.noarch
foreman-2.1.2-1.el7.noarch
```

##### APYPIE VERSION

```
Version: 0.2.2
```

##### STEPS TO REPRODUCE

ensure the centos 8 repositories have been synced.
Somehow this always results in an os called 'CentOS_Linux 8' which keeps returning when removed after every sync.
see playbook below on the reproducer

```yaml
---
- name: setup katello
  hosts: localhost
  gather_facts: true
  collections:
  - theforeman.foreman
  vars:
    katello_admin_user: admin
    katello_admin_password: redhat
    katello_server_url: 'http://katallo.linux.lab'

  tasks:
  - name: "Create operating system"
    theforeman.foreman.operatingsystem:
      username: "{{ katello_admin_user }}"
      password: "{{ katello_admin_password }}"
      server_url: "{{ katello_server_url }}"
      validate_certs: false
      name: CentOS
      os_family: Redhat
      major: '8'
      minor: '1.1911'
      password_hash: SHA256
 
  - name: "Update operating system for new release"
    theforeman.foreman.operatingsystem:
      username: "{{ katello_admin_user }}"
      password: "{{ katello_admin_password }}"
      server_url: "{{ katello_server_url }}"
      validate_certs: false
      name: CentOS
      os_family: Redhat
      major: '8'
      minor: '2.2004'
      password_hash: SHA256
 
```

##### EXPECTED RESULTS

The operating system should be updated with the new values, the same way you can do when you edit the os in the gui.

##### ACTUAL RESULTS

Using module file /home/rob/.ansible/collections/ansible_collections/theforeman/foreman/plugins/modules/installation_medium.py
Pipelining is enabled.
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 && sleep 0'
failed: [localhost] (item={'name': 'CentOS 8', 'operatingsystems': ['CentOS 8'], 'path': 'http://katello.linux.lab/pulp/repos/cloudlab/Library/custom/centos_8/CentOS_8_BaseOS_x86_64/'}) => {
    "ansible_loop_var": "installation_medium",
    "changed": false,
    "installation_medium": {
        "name": "CentOS 8",
        "operatingsystems": [
            "CentOS 8"
        ],
        "path": "http://katello.linux.lab/pulp/repos/cloudlab/Library/custom/centos_8/CentOS_8_BaseOS_x86_64/"
    },
    "invocation": {
        "module_args": {
            "locations": [
                "zolder"
            ],
            "name": "CentOS 8",
            "operatingsystems": [
                "CentOS 8"
            ],
            "organizations": [
                "Zolder Cloud Lab"
            ],
            "os_family": null,
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "path": "http://katello.linux.lab/pulp/repos/cloudlab/Library/custom/centos_8/CentOS_8_BaseOS_x86_64/",
            "server_url": "https://katello.linux.lab",
            "state": "present",
            "updated_name": null,
            "username": "admin",
            "validate_certs": false
        }
    },
    "msg": "Found too many (2) results while searching for operatingsystems with title~\"CentOS 8\""
}

```

```

### LINKED ISSUE #947: installation_medium should be able to handle an os with different versions
##### SUMMARY

Currently the installation medium can only attach to the os name and does not consider the possible versions.
ie 
if there is an os named CentOS version 8 and an os named CentOS version 7 it cannot attach to either of them due to the fact that it checks only for the os name which is CentOS in both cases, after which it will fail with an error.
It should be possible to provide versions (major and minor) so that it can decide which os it should attach to.
##### ISSUE TYPE
 - Feature Idea

### PR DESCRIPTION
**Pull request by evgeni**
Fixes: #946
Fixes: #947

### CODE REVIEW

File: plugins/doc_fragments/foreman.py [hunk 1]
```diff
@@ -273,6 +273,8 @@ class ModuleDocFragment(object):
       - Only available for Katello installations.
     required: false
     type: str
+extends_documentation_fragment:
```
Comment (evgeni) [inline plugins/doc_fragments/foreman.py]: I wonder if that even works.

Comment (evgeni) [reply plugins/doc_fragments/foreman.py]: Well, no, of course not. At least not with ansible-doc

Comment (mdellweg) [reply plugins/doc_fragments/foreman.py]: :disappointed:

File: plugins/doc_fragments/foreman.py [hunk 2]
```diff
@@ -299,3 +295,25 @@ class ModuleDocFragment(object):
...
+    type: list
+    elements: str
+'''
+
+    OPERATINGSYSTEM = '''
+options:
+  operatingsystem:
+    description:
+      - Operating systems are looked up by their title which is composed as "<name> <major>.<minor>".
+      - You can ommit the version part as long as you only have one operating system by that name.
```
Comment (mdellweg) [inline plugins/doc_fragments/foreman.py]:
```suggestion
      - You can omit the version part as long as you only have one operating system by that name.
```

File: plugins/doc_fragments/foreman.py [hunk 3]
```diff
@@ -299,3 +295,25 @@ class ModuleDocFragment(object):
...
     type: str
 '''
+
+    OPERATINGSYSTEMS = '''
+options:
+  operatingsystems:
+    description:
+      - List of operating systems the entity should be assigned to.
+      - Operating systems are looked up by their title which is composed as "<name> <major>.<minor>".
+      - You can ommit the version part as long as you only have one operating system by that name.
```
Comment (mdellweg) [inline plugins/doc_fragments/foreman.py]:
```suggestion
      - You can omit the version part as long as you only have one operating system by that name.
```

**Comment by evgeni** [comment]
I have opened https://github.com/ansible/ansible/issues/71782 and https://github.com/ansible/ansible/issues/71783, but that won't help us right now. (update: both were rejected :cry:)

**Comment by evgeni** [comment]
@RobVerduijn would the wording in this PR have saved you from the problems you had when referencing operating systems?

**Comment by RobVerduijn** [comment]
I can't find the wording you are referring to in this pull request.
The below would be nice for the
- installation_medium module
- partition_table module
```
The OS family the template shall be assigned with
When multiple matches are posible for the os name, use a more specific os name that is similar to the name in the 'title' field from the 'hammer os list' command.
```

**Comment by evgeni** [comment]
@RobVerduijn I meant the following two fragments:
https://github.com/theforeman/foreman-ansible-modules/blob/691110dfc9c68b2eb4acb1c3ec877ad45e3c66eb/plugins/doc_fragments/foreman.py#L301-L305
https://github.com/theforeman/foreman-ansible-modules/blob/691110dfc9c68b2eb4acb1c3ec877ad45e3c66eb/plugins/doc_fragments/foreman.py#L313-L316

They would be applied to the installation medium, partition table, image, os_default_template, host and hostgroup modules. (some modules refer to *multiple* operating systems, and some to a single one, thus two snippets)

**Comment by RobVerduijn** [comment]
Yes
That would have saved me quite a few hours of debugging.

**Comment by evgeni** [comment]
Thanks! And sorry for the caused confusion.

**Review by mdellweg — APPROVED** [review]

**Pull request merged**